plugins/node/opentelemetry-plugin-express/test/express.test.ts

```diff
@@ -102,6 +104,10 @@ describe('Express Plugin', () => {
       await tracer.withSpan(rootSpan, async () => {
         await httpRequest.get(`http://localhost:${port}/toto/tata`);
         rootSpan.end();
+        assert.strictEqual(
+          finishListenerCount,
+          1
+        );
```

Comment:
I'm not sure about this test, it fails without the fix, but it feels somewhat brittle. 

You can replicate the reported warning by registering the middleware above 11 times, which will print `MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 finish listeners added to [ServerResponse]. Use emitter.setMaxListeners() to increase limit`.

Reply:
Well it just make sure we are only adding a specific number of listener, one for each async middleware. 
Your changes only listen for `finish` for async middleware whereas it was for all middlewares before (my mistake) so if the user have more than 10 async middleware, it will still generate that warning.

Reply:
In this case wouldn't it be better to just use one listener and keep track of started spans? In case the listeners fires we could iterate through the list and close everything, which isn't closed yet.

Reply:
Yeah i agree that it would be much better

Reply:
After re-reading the code, I don't think we have an issue here.

According to the documentation under https://expressjs.com/en/guide/using-middleware.html:

> If the current middleware function does not end the request-response cycle, it must call next() to pass control to the next middleware function. Otherwise, the request will be left hanging.

As the `finish` listener is removed when calling `next()` and middlewares are called in serial, I don't think we can end up with more than one otel `finish` listener in this case.

Reply:
There are different possibilities:
1. sync middleware calling `next`
2. sync middleware ending the response
3. async middleware calling `next`
4. async middleware ending the reponse

With your fix, we correctly handle 1, 2 and 4. However when the middleware is doing async work before calling `next`, `spanHasEnded` is `false` since that part is run synchronously after calling the middleware, therefore its adding a listener.

I think storing spans is the only way to do this properly :/

Reply:
Sorry for the late reply, I now added tests for all cases you described, but can't reproduce the problem in any of them (also not in 3).

Reply:
@vmarchaud could you take another look if I missed anything in those test cases? Thank you 🙇

Reply:
@johanneswuerbach Looks good to me